debezium-core/src/main/java/io/debezium/connector/common/BaseSourceTask.java

```diff
@@ -131,6 +131,10 @@ public final void start(Map<String, String> props) {
 
             this.coordinator = start(config);
         }
+        catch (RetriableException e) {
+            LOGGER.warn("Stopping task due to retriable exception during start", e);
+            stop(true);
```

Comment:
> That way, Kafka Connect should restart the connector if we bubble up a `RetriableException` from `start()`?

This was my original idea but it won't. As soon as `task.start()` throws _any_ exception, it will propagate through `BaseSourceTask#start` → `WorkerSourceTask#execute` → `WorkerTask#doRun` and cause [task termination](https://github.com/apache/kafka/blob/2.7.1/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java#L189-L192):
```java
    execute();
} catch (Throwable t) {
    log.error("{} Task threw an uncaught and unrecoverable exception. Task is being killed and will not recover until manually restarted", this, t);
    throw t;
```
This is exactly the error message I saw when had originally implemented the above approach.

Reply:
> With the currently suggested approach, I'm concerned the task will be in a not fully initialized state, e.g. `coordinator` still being null.

As far as I understand, in this state, the task won't do anything except for try to restart during the next poll. If the next attempt succeeds, it will fully initialize.

Given that Kafka Connect considers any exception during start as unrecoverable, and this is by design, a proper approach might be to not have any retriable exceptions thrown during start. This would require moving all the database interaction logic (get real database name, start transaction) out of the `start` and making it "lazy". With the approach being worked in [DBZ-2975](https://issues.redhat.com/browse/DBZ-2975), the "real database part" could be even moved out of the task to the connector.